From c39c2a60f53cf67f4095d790375ef606798e40fc Mon Sep 17 00:00:00 2001
From: Hendrik Brueckner <brueckner@redhat.com>
Date: Wed, 23 Jul 2014 07:51:21 -0400
Subject: [s390] fix restore of invalid floating-point-control

Message-id: <1406101881-10995-1-git-send-email-brueckner@redhat.com>
Patchwork-id: 86312
O-Subject: [RHEL7.1 PATCH] [s390] s390: fix restore of invalid floating-point-control
Bugzilla: 1121965
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

Description
-----------
commit 9f86745722d95bc7f855069bd82285bd10dc97ff
Author: Martin Schwidefsky <schwidefsky@de.ibm.com>
Date:   Tue Jul 15 10:41:37 2014 +0200

    s390: fix restore of invalid floating-point-control

    The fixup of the inline assembly to restore the floating-point-control
    register needs to check for instruction address *after* the lfcp
    instruction as the specification and data exceptions are suppresssing.

    Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>

Bugzilla
--------
BZ 1121965
https://bugzilla.redhat.com/show_bug.cgi?id=1121965

Upstream status of the patch
----------------------------
The patch is upstream as of kernel version
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=9f86745722d95bc7f855069bd82285bd10dc97ff

Brew
----
https://brewweb.devel.redhat.com/taskinfo?taskID=7721888

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/s390/include/asm/switch_to.h b/arch/s390/include/asm/switch_to.h
index 0f86582..e954d63 100644
--- a/arch/s390/include/asm/switch_to.h
+++ b/arch/s390/include/asm/switch_to.h
@@ -50,8 +50,8 @@ static inline int restore_fp_ctl(u32 *fpc)
   return 0;
 
  asm volatile(
-  "0: lfpc    %1\n"
-  " la %0,0\n"
+  " lfpc    %1\n"
+  "0: la %0,0\n"
   "1:\n"
   EX_TABLE(0b,1b)
   : "=d" (rc) : "Q" (*fpc), "0" (-EINVAL));
-- 
1.7.1